### ISSUE TITLE
videobuf_vmalloc Deprecated

### ISSUE THREAD

**Issue by PseudoResonance** [stoth68000/sc0710#17]
Videobuf has long been deprecated, and now Arch kernel 6.4 has removed the videobuf_vmalloc module that this project depends on. [CONFIG_VIDEOBUF_VMALLOC=m on line 5844](https://gitlab.archlinux.org/archlinux/packaging/packages/linux/-/commit/50de1591172a64cb6fac2820e9653a179fd8a230#dfba7aade0868074c2861c98e2a9a92f3178a51b_5844_5864)

**Comment by NuiceT** [comment]
I just noticed when trying to run `make load`:

```bash
root@elgato:~/sc0710# make load
make -C /lib/modules/5.15.0-78-generic/build M=/root/sc0710 modules
make[1]: Entering directory '/usr/src/linux-headers-5.15.0-78-generic'
make[1]: Leaving directory '/usr/src/linux-headers-5.15.0-78-generic'
sudo dmesg -c >/dev/null
sudo cp /dev/null /var/log/debug
#sudo modprobe videobuf2-core
sudo modprobe videobuf2-common
sudo modprobe videodev
#sudo modprobe videobuf-dma-sg
sudo modprobe videobuf-vmalloc
sudo insmod ./sc0710.ko \
	thread_dma_poll_interval_ms=2 \
	dma_status=0
insmod: ERROR: could not insert module ./sc0710.ko: Unknown symbol in module
make: *** [Makefile:26: load] Error 1
```

**Comment by pborrell** [comment]
I'm hoping someone can beat me to forcing this to upgrade as I only have maybe an hour a week to devote to hacking drivers in Linux. But I'm looking at this as a challenge accepted moment.

**Comment by ApolloXI14** [comment]
Try changing `videobuf-vmalloc` to `videobuf2-vmalloc`

**Comment by PseudoResonance** [comment]
That's not how this works unfortunately. Videobuf2 and videobuf are incompatible and you can't just quickly swap it out. It requires a not insignificant amount of rewriting the driver. [Here's](https://lwn.net/Articles/447435/) something that describes a few of these differences.

**Comment by stoth68000** [comment]
I have a lot going on with work, I can’t look at this anytime soon. This is
one of many pet projects for me, just the reality of life.

I acknowledge the ongoing discussion, I’m just in no position to work on
it. If you need the driver to work, use a kernel it was developed against.
My last update added support for a recent Ubuntu distribution, detail are
probably in the commits.

- Steve

**Comment by pborrell** [comment]
> Try changing `videobuf-vmalloc` to `videobuf2-vmalloc`

I looked at that, and based on the structures,  of 1 and 2
![image](https://github.com/stoth68000/sc0710/assets/14113918/2484505f-a1bc-4945-b990-cae4a10d5f63)
![image](https://github.com/stoth68000/sc0710/assets/14113918/56fbafd2-f4fa-4940-a20c-32effe01834c)

They aren't drop in replacements. Steve, I feel where you're coming from with the lot of work going on if I had 120 hours to devote to this I could most likely update it myself, but with the changes to the queue system alone. With small amounts of work here and there, it would most likely take me a few months to move the queue to the non-deprecated system, let alone the malloc and the buffer.

**Comment by stoth68000** [comment]
>
>
> Steve, I feel where you're coming from with the lot of work going on if I
> had 120 hours to devote to this I could most likely update it myself, but
> with the changes to the queue system alone. With small amounts of work here
> and there, it would most likely take me a few months to move the queue to
> the non-deprecated system, let alone the malloc and the buffer.
>
>
>
Indeed, it's just volunteer work, very low priority for me right now.

We do what we can - when we can.

-- 
Steven Toth - Kernel Labs
http://www.kernellabs.com
+1.646.355.8490

**Comment by pborrell** [comment]
I'm looking through the headers and the c code and at least it seems that most of it is already using video for linux 2 implementations. So it's not a full rewrite.

**Comment by ApolloXI14** [comment]
Thanks for the answer and article Josh and looking further into this Peter. I'm new to V4L and kernel modules myself and wanted to help since I just started using this Elgato and am happy with it, but I am disappointed in the lack of official Linux support, which encouraged me to start digging myself lately, which brought me here. Impressed and grateful for what you've already done Steve.

**Comment by PseudoResonance** [comment]
If you want, here's an example of a rewrite I found. [sh_vou videobuf2 rewrite](https://github.com/analogdevicesinc/linux/commit/57af3ad59d953f300a1fcb143e72d024aff73550). Kinda gives an idea of the work required. Doesn't look too bad, but would also require a decent understanding of the kernel and V4L2, which I unfortunately don't know much about. Maybe if I have time...

**Comment by stephematician** [comment]
I'll have a crack at this as I've been working on https://github.com/v4l2loopback/v4l2loopback and I need to learn a bit about the `videobuf2` framework for that, too. So this is a good project to start with.

**Comment by VixiKitsune** [comment]
> I'll have a crack at this as I've been working on https://github.com/v4l2loopback/v4l2loopback and I need to learn a bit about the `videobuf2` framework for that, too. So this is a good project to start with.

I had the same idea a few weeks ago. I found there were a lot of equivalent vb2 functions to what's used in this driver. I swapped to them but didn't get round to testing as I've been busy. Though it's not really relevant anymore with Nakildia's fork.

**Comment by stephematician** [comment]
@VixiKitsune - thanks, yes that fork (sort of) was just brought to my attention. Saves me the trouble. For anyone else who is looking for an update to this project that works on kernels 6.12 and up: see https://github.com/Nakildias/sc0710.
